### PR TITLE
first step of constraint transformation

### PR DESCRIPTION
**Pull request by migeed-z**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80147
* #80110
* __->__ #80102
* #80095

Differential Revision: [D37387412](https://our.internmc.facebook.com/intern/diff/D37387412)

Here, we transform the high level constraints to lower level constraints. For example we can expand a precision constraint into a set of equality constraints. The same thing for consistency constraints. We do not yet transform them to Z3. This will be in the next step.

### CODE REVIEW

**Comment by migeed-z** [comment]
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37387412).

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py [hunk 1]
```diff
@@ -0,0 +1,724 @@
...
+
+    elif isinstance(constraint, CanReshape):
+        transformed, counter = transform_can_reshape(constraint, counter)
+        return transformed, counter
+
+    elif isinstance(constraint, ApplyBroadcasting):
+        transformed, counter = transform_apply_broadcasting(constraint, counter)
+        return transformed, counter
+    else:
+        return constraint, counter
```
Comment (jamesr66a) [inline torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:145]: Can we turn this if/else tree into a registration pattern like in `constraint_generator.py`? It's also kind-of strange to be directly dispatching to another function for half of the constraint types (`CalcConv` etc) but then have logic embedded in the if/else for others

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+
+
+def transform_conv(constraint, counter):
+    """
+    :param constraint: Calc-conv
+    :return: new counter and the transformed constraint
```
Comment (jamesr66a) [inline torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:151]: nit: in the PT codebase we use the google docstring style

PT docs: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#writing-documentation
google docstring style: https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py [hunk 3]
```diff
@@ -0,0 +1,724 @@
...
+        final_tensor_4_constraint_no_padding,
+        final_tensor_4_constraint_padding_arg1,
+        final_tensor_4_constraint_padding_arg2
+    ])
+
+    return Conj([final_result, *nat_dims_1, *nat_dims_2, *nat_dims_3, *nat_dims_4]), counter
+    # return final_result, counter
+
+
+def generate_all_possibilities(my_list):
```
Comment (jamesr66a) [inline torch/fx/experimental/migrate_gradual_types/constraint_transformation.py:306]: Can you name this something more specific? It's not clear what this does from the name without reading the body of the function

**Review by jamesr66a — COMMENTED** [review]
This looks fine from a coding perspective, but maybe you could walk us all through what's going on at some point. It would help me (even though I've been working with you on this for a while) and I definitely think it would help @anijain2305  and @jansel

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py [hunk 4]
```diff
@@ -0,0 +1,724 @@
...
+from torch.fx.experimental.migrate_gradual_types.constraint import ApplyBroadcasting, Prod, F
+from torch.fx.experimental.migrate_gradual_types.operation import op_eq, op_precision, op_leq, op_matching
+from torch.fx.experimental.migrate_gradual_types.operation import op_consistency, op_neq
+from torch.fx.experimental.migrate_gradual_types.operation import op_mul, op_add, op_sub, op_div, op_mod
+from torch.fx.experimental.migrate_gradual_types.util import gen_tensor_dims, gen_nat_constraints, gen_dvar
+from torch.fx.tensor_type import TensorType, Dyn
+
+
+# run this as the first transformation step till you get a fixed point
+# handles: consistency, matching, precision, GLB, leq, broadcasting, reshape, convolution..
```
[the remark this answers is not in the document]
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_transformation.py]: done

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py (hunk 2 above)
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_transformation.py]: done

File: torch/fx/experimental/migrate_gradual_types/constraint_transformation.py (hunk 3 above)
Comment (migeed-z) [reply torch/fx/experimental/migrate_gradual_types/constraint_transformation.py]: done

**Review by anijain2305 — APPROVED** [review]

**Review by jamesr66a — APPROVED** [review]

**Pull request closed without merging**